webpack/scenes/ContentViews/Details/Promote/ContentViewVersionPromote.js

```diff
@@ -0,0 +1,165 @@
+      environment_ids: userCheckedItems.map(item => item.id),
+      force: true,
+    }));
+  };
+
+  useEffect(() => {
+    if (promoteResolved && promoteResponse) {
+      setIsOpen(false);
+      dispatch(getContentViewVersions(cvId));
+    }
+    if (promoteError) {
+      setSaving(false);
+    }
+  }, [promoteResponse, promoteResolved, promoteError, setSaving, setIsOpen, dispatch, cvId]);
+
+  const envPathFlat = useMemo(() => {
+    if (!environmentPathLoading) {
+      const { results } = environmentPathResponse || {};
+      return results.map(result => result.environments).flatten();
+    }
+    return [];
+  }, [environmentPathResponse, environmentPathLoading]);
+
+  const prior = useCallback(
+    env => envPathFlat.find(item => item.id === env.prior.id),
+    [envPathFlat],
+  );
+
+  const isChecked = useCallback(
+    env => (userCheckedItems.includes(env) ||
+      versionEnvironments.filter(item => item.id === env.id).length),
+    [userCheckedItems, versionEnvironments],
+  );
+
+  const isValid = useCallback((env) => {
+    if (!env.prior) return true;
+    if (!isChecked(prior(env))) return false;
+    return isValid(prior(env));
+  }, [prior, isChecked]);
+
+  useEffect(() => {
+    setForcePromote(userCheckedItems.filter(item => !isValid(item)));
+  }, [userCheckedItems, setForcePromote, isValid]);
+
+  return (
+    <Modal
+      title={__(`Promote version ${versionNameToPromote}`)}
+      isOpen={show}
+      variant={ModalVariant.large}
+      onClose={() => {
+        setIsOpen(false);
+      }}
+      appendTo={document.body}
+    >
+      {saving && <Loading loadingText={__('Please wait while task starts..')} />}
+      {!saving &&
+        <>
+          <Form>
+            <FormGroup label="Description" fieldId="description">
+              <TextArea
+                isRequired
+                type="text"
+                id="description"
+                aria-label="input_description"
+                name="description"
+                value={description}
+                onChange={value => setDescription(value)}
+              />
+            </FormGroup>
+          </Form>
+          <hr />
+          {forcePromote.length > 0 && (
+          <Alert variant="info" isInline title="Force Promotion">
+            <p>{__('Selected environments are out of the environment order. ' +
+              'The recommended practice is to promote to the next environment in the path.')}
+            </p>
+            <ComponentEnvironments environments={forcePromote} />
+          </Alert>)}
+          <EnvironmentPaths
+            userCheckedItems={userCheckedItems}
+            setUserCheckedItems={setUserCheckedItems}
+            promotedEnvironments={versionEnvironments}
+            publishing={false}
+          />
+          <Form>
+            <ActionGroup>
+              <Button
+                aria-label="promote_content_view"
+                variant="primary"
+                isDisabled={saving}
```

Comment:
I'm able to press the Promote button even if the version has been promoted to all available environments. It seems like that should be disabled and perhaps an 'empty state' could be applied to this component "This version is already published to all available environments" or something like that

Reply:
+1 This was discussed, I think we decided to add this view to a separate ticket as not to bloat this one? @sjha4

Reply:
Totally works for me as long as there's a Redmine filed for it

Reply:
This would be a nice to have addition. Opened https://projects.theforeman.org/issues/33275 for this.